### PR TITLE
fix(event_handler): raise more specific SerializationError exception for unsupported types in data validation

### LINKED ISSUE #4300: Bug: APIGatewayHttpResolver fails with ValueError: [TypeError('cannot convert dictionary update sequence element #0 to a sequence'), TypeError('vars() argument must have __dict__ attribute')]
### Expected Behaviour

I expect the APIGatewayHttpResolver to route my request to the proper annotated function, eg:
```
@app.get("/something")
def something():
    # does something...
```

### Current Behaviour

For some requests, I get the error `ValueError: [TypeError('cannot convert dictionary update sequence element #0 to a sequence'), TypeError('vars() argument must have __dict__ attribute')]`

I'm using the serverless framework to package and deploy my app. I'm also using the [serverless-python-requirements](https://github.com/serverless/serverless-python-requirements) plugin to package my app.

See attached snippet for full traceback.

### Code snippet

```python
ANY /teams (λ: handler)
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/openapi/encoders.py", line 272, in _dump_other
backend-1  |
    data = dict(obj)
           ^^^^^^^^^
TypeError: cannot convert dictionary update sequence element #0 to a sequence
backend-1  |
During handling of the above exception, another exception occurred:
backend-1  |
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/openapi/encoders.py", line 276, in _dump_other
    data = vars(obj)
           ^^^^^^^^^
TypeError: vars() argument must have __dict__ attribute
backend-1  |
The above exception was the direct cause of the following exception:
backend-1  |
Traceback (most recent call last):
  File "/app/node_modules/serverless-offline/src/lambda/handler-runner/python-runner/invoke.py", line 97, in <module>
backend-1  |
    result = handler(input['event'], context)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/tracing/tracer.py", line 317, in decorate
    response = lambda_handler(event, context, **kwargs)
               ^^^^^^^^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/app/handlers.py", line 192, in handler
    return app.resolve(event, context)
           ^^^^^^^^^^^^^^^^^^^
^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/api_gateway.py", line 1918, in resolve
backend-1  |
    response = self._resolve().build(self.current_event, self._cors)
               ^^^^^^^^^^^^^^
^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/api_gateway.py", line 2025, in _resolve
backend-1  |
    return self._call_route(route, route_keys)  # pass fn args
backend-1  |
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/api_gateway.py", line 2103, in _call_route
backend-1  |
    route(router_middlewares=self._router_middlewares, app=self, route_arguments=route_arguments),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/api_gateway.py", line 407, in __call__
backend-1  |
    return self._middleware_stack(app)
backend-1  |
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/api_gateway.py", line 1314, in __call__
backend-1  |
    return self.current_middleware(app, self.next_middleware)
backend-1  |
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/middlewares/base.py", line 121, in __call__
    return self.handler(app, next_middleware)
backend-1  |
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/middlewares/openapi_validation.py", line 133, in handler
    return self._handle_response(route=route, response=response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/middlewares/openapi_validation.py", line 140, in _handle_response
    response.body = self._serialize_response(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/middlewares/openapi_validation.py", line 200, in _serialize_response
    return jsonable_encoder(response_content, custom_serializer=self._validation_serializer)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/openapi/encoders.py", line 122, in jsonable_encoder
backend-1  |
    return _dump_sequence(
           ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/openapi/encoders.py", line 245, in _dump_sequence
    jsonable_encoder(
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/openapi/encoders.py", line 145, in jsonable_encoder
    return _dump_other(
           ^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/openapi/encoders.py", line 279, in _dump_other
    raise ValueError(errors) from e
ValueError: [TypeError('cannot convert dictionary update sequence element #0 to a sequence'), TypeError('vars() argument must have __dict__ attribute')]
```
```


### Possible Solution

_No response_

### Steps to Reproduce

My project is not easy to reproduce locally and I can't share it publicly, but I would be OK with sharing my screen or sending certain files privately to powertools-lambda-python contributors.

### Powertools for AWS Lambda (Python) version

latest

### AWS Lambda function runtime

3.11

### Packaging format used

PyPi

### Debugging logs

_No response_

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #4300 

### Changes

The current implementation of the Data Validation does not natively support serializing some types of objects, including SQLAlchemy model objects and other custom classes that may not have a default serialization mechanism. This can lead to issues when trying to return these objects as the error is confuse.

In this PR, we are adding more robust exception handling and a clear message for cases where the object to be serialized is not supported.

Example of this implementation:

```python
from typing import Any
from json import JSONEncoder
from dataclasses import is_dataclass, asdict
from sqlalchemy.orm.base import InstrumentedAttribute
from pydantic.main import ModelMetaclass

class UnsupportedObjectEncoder(JSONEncoder):
    def default(self, obj: Any) -> Any:
        if isinstance(obj, InstrumentedAttribute):
            return str(obj)
        elif is_dataclass(obj):
            return asdict(obj)
        elif isinstance(obj, ModelMetaclass):
            return obj.dict()
        else:
            try:
                return super().default(obj)
            except TypeError:
                # Handle unsupported objects by returning a default representation
                return str(obj)

def custom_serializer(obj: Any) -> str:
    return json.dumps(obj, separators=(",", ":"), cls=UnsupportedObjectEncoder)

app = APIGatewayRestResolver(enable_validation=True, serializer=custom_serializer)
```

### User experience

Before this implementation, the customer received an incomprehensible error like:

```sh
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/openapi/encoders.py", line 272, in _dump_other
backend-1  |
    data = dict(obj)
           ^^^^^^^^^
TypeError: cannot convert dictionary update sequence element #0 to a sequence
backend-1  |
During handling of the above exception, another exception occurred:
backend-1  |
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aws_lambda_powertools/event_handler/openapi/encoders.py", line 276, in _dump_other
    data = vars(obj)
```

Now we've added a clear message in the exception

`[ERROR] EncoderError: ('Unable to serializer the object <app.MyClass object at 0x7ffff27f45c0> as it is not a supported type. Error details: [TypeError("\'MyClass\' object is not iterable"), TypeError(\'vars() argument must have __dict__ attribute\')]', 'See: https://docs.powertools.aws.dev/lambda/python/latest/core/event_handler/api_gateway/#serializing-objects')
`
## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: docs/core/event_handler/api_gateway.md
Comment (sthulb) [inline docs/core/event_handler/api_gateway.md]: I think I'd rather have this as an exhaustive list, it'll improve readability and stop people from having to read the code

File: aws_lambda_powertools/event_handler/openapi/encoders.py
Comment (sthulb) [inline aws_lambda_powertools/event_handler/openapi/encoders.py]: This might be better as `SerializationError` or `SerDeError` if we ever plan to decode too

File: docs/core/event_handler/api_gateway.md
Comment (leandrodamascena) [reply docs/core/event_handler/api_gateway.md]: I'm not happy about it either, but what's the best way to make that explicit? This is important to let customers now the most popular types we can serializer. 
I also don't like adding a link to the code because we can change the code (add/remove lines) and the link will point somewhere that doesn't make sense.

Any idea?

File: aws_lambda_powertools/event_handler/openapi/encoders.py
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/openapi/encoders.py]: Great catch @sthulb! I've changed the name to `SerializationError`

File: aws_lambda_powertools/event_handler/openapi/encoders.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/openapi/encoders.py:108]: suggestion: take a look at `singledispatch` to create a dispatcher to make this more maintainable, as it grows with the different types we're handling it.

File: docs/core/event_handler/api_gateway.md
Comment (sthulb) [reply docs/core/event_handler/api_gateway.md]: I think having an actual list of the types you have above is enough, but actually presented as a list.

File: aws_lambda_powertools/event_handler/openapi/encoders.py
Comment (heitorlessa) [reply aws_lambda_powertools/event_handler/openapi/encoders.py:108]: (seeing on laptop now) - it sounds like those were already there, so disregard it. 

it's easier to have an issue to refactor these parts

File: aws_lambda_powertools/event_handler/openapi/encoders.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/openapi/encoders.py]:
```suggestion
        raise SerializationError(
            f"Unable to serialize the object {obj} as it is not a supported type. Error details: {exc}",
            "See: https://docs.powertools.aws.dev/lambda/python/latest/core/event_handler/api_gateway/#serializing-objects",
```

fix: typo and double serialization.

tip: when using `f-string` you don't need `str(obj)` otherwise you're doing double serialization to str.

File: docs/core/event_handler/api_gateway.md
Comment (heitorlessa) [reply docs/core/event_handler/api_gateway.md]:
```suggestion
With data validation enabled, we natively support serializing the following data types to JSON:

Data type | Serialized type
------------------------------------------------- | ---------------------------------------------------------------------------------
**Pydantic models** | `dict`
**Python Dataclasses** | `dict`
**Enum** | Enum values
**Datetime** | Datetime ISO format string
**Decimal** | `int` if no exponent, or `float`
**Path** | `str`
**UUID** | `str`
**Set** | `list`
**Python primitives** _(dict, string, sequences, numbers, booleans)_ | [Python's default JSON serializable types](https://docs.python.org/3/library/json.html#encoders-and-decoders){target="_blank" rel="nofollow"}

???+ info "See [custom serializer section](#custom-serializer) for bringing your own."
    Otherwise, we will raise `SerializationError` for any unsupported types _e.g., SQLAlchemy models_.
```

editing: **1/** moved into a table, **2/** removed non-permalink to our encoder src, **3/** linked custom serializer, and **4/** added a note about `SerializationError`

File: docs/core/event_handler/api_gateway.md
Comment (heitorlessa) [inline docs/core/event_handler/api_gateway.md]: merged with the suggestion but GH doesn't allow me to extend it.

File: tests/functional/event_handler/test_openapi_encoders.py
Comment (heitorlessa) [inline tests/functional/event_handler/test_openapi_encoders.py]:
```suggestion
    with pytest.raises(SerializationError, match="Unable to serialize the object*"):
```

typo in exception and assertion

**Review by heitorlessa — CHANGES_REQUESTED** [review]
suggested changes in doc to make this more visible and easier to parse, and a note about encoders refactoring to `singledispatch`. In the ideal world, most of this logic would be shared this level of serialization would benefit other areas like `idempotency` to cut duplication.

File: tests/functional/event_handler/test_openapi_encoders.py
Comment (leandrodamascena) [reply tests/functional/event_handler/test_openapi_encoders.py]: Done

File: docs/core/event_handler/api_gateway.md
Comment (leandrodamascena) [reply docs/core/event_handler/api_gateway.md]: Done

File: aws_lambda_powertools/event_handler/openapi/encoders.py
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/openapi/encoders.py:108]: I'll open an issue to refactor that.

**Comment by leandrodamascena** [comment]
Ready to review again @sthulb and @heitorlessa

**Comment by heitorlessa** [comment]
looking...

File: docs/core/event_handler/api_gateway.md
Comment (heitorlessa) [inline docs/core/event_handler/api_gateway.md]:
```suggestion
#### Supported types for response serialization
```

rename: aligning section with the content.

I've ran the docs locally and this section was confusing me with `Custom serializer`, and whether this was about request serialization. 

I'd suggest changing the name to match the table content and our intent (use case driven).

**Review by heitorlessa — CHANGES_REQUESTED** [review]
Last comment before merging - changing the docs section name.

File: docs/core/event_handler/api_gateway.md
Comment (leandrodamascena) [reply docs/core/event_handler/api_gateway.md]: Done

**Comment by heitorlessa** [comment]
LGTM. @sthulb I'm gonna resolve the merge conflict and get your approval to merge, as @leandrodamascena is on public holiday today

**Update**: nvm, he's working on it

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**
